source/common/filter/http/filter_config_discovery_impl.cc

```diff
@@ -148,12 +164,20 @@ void FilterConfigSubscription::onConfigUpdate(
     const std::vector<Config::DecodedResourceRef>& added_resources,
     const Protobuf::RepeatedPtrField<std::string>& removed_resources, const std::string&) {
   if (!removed_resources.empty()) {
-    ENVOY_LOG(error,
-              "Server sent a delta ExtensionConfigDS update attempting to remove a resource (name: "
-              "{}). Ignoring.",
-              removed_resources[0]);
-  }
-  if (!added_resources.empty()) {
+    ASSERT(removed_resources.size() == 1);
+    ENVOY_LOG(debug, "Removing filter config {}", filter_config_name_);
+    const auto pending_update = std::make_shared<std::atomic<uint64_t>>(
+        (factory_context_.admin().concurrency() + 1) * filter_config_providers_.size());
+    for (auto* provider : filter_config_providers_) {
+      provider->onConfigRemoved([this, pending_update]() {
+        if (--(*pending_update) == 0) {
+          stats_.config_reload_.inc();
+        }
+      });
+    }
+
+    last_config_hash_ = 0;
+  } else if (!added_resources.empty()) {
```

Comment:
qq: Can resources be added and removed at the same time?

Reply:
I think we'll never get both a removal and addition/update of the same resource and we only subscribe to one resource here so we shouldn't ever have to handle both at the same time. It would be nice to have this be better encoded in the API, but I'm not sure if there's a great way to handle this besides ASSERTs? cc @htuch

Reply:
Yeah, I think the GrpcMux should ensure this property, but I can't say for certainty if it's true. We should make it part of the API IMHO, i.e. if the server does do a simultaneous add/remove, we don't expose any of the subscriptions to this. CC @dmitri-d @adisuissa

Reply:
In the unified mux implementation there's a check for this: https://github.com/envoyproxy/envoy/pull/15473/files#diff-4228bc8f4213ff904d2a3584f69a2cf049180957f668d8bc0125d9bfa3379fa2R119. If failed, the DeltaDiscoveryResponse is discarded.